src/Sylius/Bundle/AddressingBundle/spec/Validator/Constraints/ProvinceAddressConstraintValidatorSpec.php

```diff
@@ -62,6 +72,32 @@ function it_does_not_add_violation_because_a_violation_exists(
         $this->validate($address, $constraint);
     }
 
+    function it_does_not_add_violation_because_a_violation_exists_when_address_is_the_root_object(
+        RepositoryInterface $countryRepository,
+        AddressInterface $address,
+        Country $country,
+        ProvinceAddressConstraint $constraint,
+        ExecutionContextInterface $context
+    ): void {
+        $country->getCode()->willReturn('PL');
+        $address->getCountryCode()->willReturn('PL');
+        $countryRepository->findOneBy(['code' => 'PL'])->willReturn($country);
+
+        $country->hasProvinces()->willReturn(true);
+        $address->getProvinceCode()->willReturn(null);
```

Comment:
I believe this and the similar mocks in the spec above are not needed? These functions are never called as the `isProvinceValid` function is not reached. I even removed them and specs were still green 💃

Reply:
I think I suggested to add those, but I can't remember why.
Were the tests false positive, @rimas-kudelis?

Reply:
@Zales0123 Yes, I think that was it... that's how we tested that we need the fix.
With the fix, the builds are green either way, but without the fix, it only fails with the proper setup (the mocks).
The mocks are required to prevent regression.

Reply:
@rimas-kudelis maybe leave a comment there, because in another month who knows if anyone can remember this. 😅

Reply:
I actually put that info in the commit message. But sure, I can put it in a comment as well.

Reply:
Right. I didn't see it because GitHub hides everything after the first line and you have to click a button to see it. Thanks!

Reply:
I've added a comment to both specs.

Reply:
I guess Mateusz was pointing out just the `$address->getProvinceCode()->willReturn(null);` line, as returning null is default phpspec behvaiour

Reply:
Or after the second reading - shouldn't we rather take usage of `->shouldNotBeCalled()`?

Reply:
@lchrusciel You should read it again, I think you missed the point. It would be easier to understand if you would apply the spec changes without the fix and run phpspec.
Spoiler: it will fail, which is precisely the goal of the tests. Without those changes in the spec, it will pass.

Reply:
@vvasiloi still, `shouldNotBeCalled()` could be a valid way out too.
But I'm wondering if that wouldn't be an assertion of an implementation detail.

Reply:
Yes, that's an implementation detail, but it's not even what is tested here.
Usually,  methods like `shouldNotBeCalled`, which are asserts for behavior, are used for the test subject, which in this case is the validator class.
The address and it's related objects are the input in this case, which is how the test should be set up for failure **before** writing the implementation.
In this case, originally, it was probably written **after** the implementation, so it never had the proper input setup.
The test was not useful from the start, because it passes both with and without the implementation.
That's why TDD is about writing tests first, see that it fails, write the implementation to pass the test.

Reply:
@rimas-kudelis I checked it out again and I see that `it_does_not_add_violation_because_a_violation_exists_when_address_is_the_root_object` passes with and without your fix. 🤔 
How did this happen?

Reply:
@vvasiloi I get you point, but `$address->getProvinceCode()->willReturn(null);` should not change anything, as this is default behaviour. On the other hand, keeping it won't change anything so we can leave it as is.

Still it probably requires manual testing from our side

Reply:
@lchrusciel I added `->willReturn(null)` just to make it more explicit that stuff would not match. That line is certainly not necessary for the test, I won't deny that. :)

Reply:
Hey folks, I've tried to extract bloated changes to separate class, so it would be easier for me to comprehend what is happening here. You can find something here: https://github.com/Sylius/Sylius/pull/13599. My findings are as follows:
1. "passing empty needle to strpos()" triggers php warning, should be omitted. Therefore the bug is present
2. Since #13562 in branch 1.11 we are using string_starts_with(), so it is fix only for 1.10
3. There is no value in the proposed tests in my opinion. I'm willing to merge just one line fix in validator if we will remove proposed specs.
4. If we want to move forward, some refactor of this logic is a musthave. Amount of different paths here is just hilarious. I'm not insisting on my solution (especially that bc layer may be controversial), but for sure we should do something about it if we want to touch this class more. Also, the refactor should probably be merge to master branch.

Reply:
@lchrusciel are you saying the validator itself should be refactored, or were you talking about the tests?
The validator ir just 92 lines – that doesn't sound too bloated to me (although I didn't try to analyze its logic now).

Reply:
Also, regarding the tests, as mentioned above, we saw that current tests were passing regardless of whether or not the functionality in question worked, so we tried extending them to not pass that easy. And the special cases for `address_is_the_root_object` were added because this is a fix for a bug which only manifests itself when the address is the root object.

Reply:
Class is short, but have 10 if statements and it makes writing test quite painful. Also, I get the reason of your test - but they don't change anything. Tests are green no matter if this fix is added or not.

Reply:
![image](https://user-images.githubusercontent.com/6213903/152563409-48f4e802-9908-4504-837c-52bacdde885a.png)

Reply:
and if you comment out that `foreach` block?

Reply:
Ok, so I've tested it on PHP8.0, but the warning is triggered on PHP7.4 http://sandbox.onlinephpfunctions.com/code/ee553ff2e53d1ea3b454dea89f4dda61fdbfa3c3

Reply:
> Ok, so I've tested it on PHP8.0, but the warning is triggered on PHP7.4

OMG, and this change is not even documented by PHP!

Still, what happens if you comment out the `foreach`?

Reply:
Let me come back to the discussion 😄 As far as I see, the warning about empty string indeed happens only on the PHP 7.4. I think it was one of the roots of the misunderstanding there 🚀 

<img width="987" alt="Zrzut ekranu 2022-02-4 o 17 21 48" src="https://user-images.githubusercontent.com/6212718/152569126-4350b0a4-e494-48fa-81d6-416253044941.png">

And **indeed** when the foreach is commented out, the specs are still green. However, as far as I see, in these tests that you're changing we can call `$address->getCountryCode()->shouldNotBeCalled();` instead of only `$context->addViolation(Argument::any())->shouldNotBeCalled();` so it ensures us the `isProvinceValid` is not used nor accessed. 

```php
    function it_does_not_add_violation_because_a_violation_exists_when_address_is_the_root_object(
        RepositoryInterface $countryRepository,
        AddressInterface $address,
        Country $country,
        ProvinceAddressConstraint $constraint,
        ExecutionContextInterface $context
    ): void {
        $this->initialize($context);

        $context->getPropertyPath()->willReturn('');
        $context->getViolations()->willReturn(new \ArrayIterator([
            $this->createViolation('property_path'),
        ]));

        $address->getCountryCode()->shouldNotBeCalled();
        $context->addViolation(Argument::any())->shouldNotBeCalled();

        $this->validate($address, $constraint);
    }
```

And even though it's not **strictly** connected with the proposed fix, it's fair enough to propose both of them together in one PR 🖖 What do you think about it @rimas-kudelis @lchrusciel?

Reply:
Those extra bits of setup that were added for both tests are for the whole foreach, not just for the condition that Rimas added, which can be reproduced with PHP 7.4.
If you revert everything in changed in this PR, or just start with a 1.10 branch and remove that foreach, then the tests will still pass. Somehow I forgot to mention this in my previous comments.

Reply:
Also, why are you not using https://3v4l.org/, @lchrusciel?

Reply:
You can see the result for all versions at once there: https://3v4l.org/5ma7S

Reply:
@Zales0123 like I said, I agree that asserting that `->getCountryCode()` must not be called would make the tests fail. But it feels to me like an assertion on **How the validator works** instead of **What the expected result is**. The way I wrote it allows the test subject to remain more of a blackbox, doesn't it?

Reply:
Sorry, I missed the Mateusz's latest message.
I second Rimas here and I also said something among those line earlier.
We're testing the validator here and the test is called `it_does_not_add_violation`, so I don't see why we would assert the behavior on the address object instead of the validator context.
There's a good synergy between: `it_does_not_add_violation` and `$context->addViolation(Argument::any())->shouldNotBeCalled();`